### PR TITLE
added rh_sso cli tests, refactored some common code

### PR DESCRIPTION
**Pull request by omkarkhatavkar**
### PR Description
1. Added RH-SSO-Satellite CLI integration tests.
2. Refactored the RH-SSO utils and common utils files.
3. Added support for the Pexpect python module. 

### Test Result 
```
============================= test session starts ==============================
platform linux -- Python 3.6.8, pytest-4.6.3, py-1.8.1, pluggy-0.13.1
shared_function enabled - OFF - scope:  - storage: file
rootdir: /home/okhatavk/Satellite/robottelo
plugins: forked-1.1.3, xdist-1.31.0, cov-2.8.1, mock-1.10.4, services-1.3.12020-08-11 10:43:49 - conftest - DEBUG - Collected 3 test cases
collected 3 items

test_ldapauthsource.py                                                [100%]

=============================== warnings summary ===============================
```
### Dependant Airgun PR 
https://github.com/SatelliteQE/airgun/pull/503

### CODE REVIEW

**Comment by mirekdlugosz** [comment]
@omkarkhatavkar also, there is conflict that prevents merge of this PR. Please resolve it.

File: robottelo/rhsso_utils.py
Comment (omkarkhatavkar) [reply robottelo/rhsso_utils.py:168]: I tried that it was not working, I know robottelo.ssh is fast but it was not working with this command. I gave it a lot of try. I was getting below error hence I moved to pxssh

`E           robottelo.cli.base.CLIReturnCodeError: CLIReturnCodeError(return_code=70, stderr='Error: No such device or address @ rb_sysopen - /dev/tty\n', msg="Failed to run the command : echo '26defc5b-2de3-4b05-83f5-dfe0d17c6054.4bc20789-d4a0-46a2-ab5c-22c05ec46f6f.77bad924-e252-4daa-a7c9-9697d05339eb' | hammer auth login oauth --oidc-token-endpoint https://rhsso.redhat.com/auth/realms/satqe/protocol/openid-connect/token --oidc-authorization-endpoint https://rhsso.redhat.com/auth/realms/satqe/protocol/openid-connect/auth --oidc-client-id satellite.redhat.com-foreman-openidc --oidc-redirect-uri 'urn:ietf:wg:oauth:2.0:oob' --two-factor "   `

File: robottelo/rhsso_utils.py
Comment (omkarkhatavkar) [reply robottelo/rhsso_utils.py:131]: That more clean and understandable way

File: tests/foreman/cli/test_ldapauthsource.py
Comment (omkarkhatavkar) [reply tests/foreman/cli/test_ldapauthsource.py:341]: That was my first idea, I dont remember what I did after that. Let me get back to you on this.

File: pytest_fixtures/satellite_auth.py
Comment (mshriver) [inline pytest_fixtures/satellite_auth.py:163]: If you're going to define a function for cleanup, you can decorate that function instead of having to then explicitly call it here.

```
@request.addfinalizer
def _cleanup():
    update_rhsso_settings_in_satellite(revert=True)
```

That being said, for such a simple call, it could be done via a lambda passed to `request.addfinalizer`.

And _that_ being said, here I would just yield. Its my preferred method for doing fixture cleanup when I'm not concerned about runtime exceptions in a complex fixture.

```
@pytest.fixture
def rhsso_settings():
    update_rhsso_settings_in_satellite()
    yield
     update_rhsso_settings_in_satellite(revert=True)
```

File: pytest_fixtures/satellite_auth.py
Comment (mshriver) [inline pytest_fixtures/satellite_auth.py:177]: Again consider yielding and then calling your two statements for cleanup.

File: robottelo/api/utils.py
Comment (mshriver) [inline robottelo/api/utils.py:985]: Consider flipping these blocks instead of inverting the logic.

```
if revert:
    <do the revert stuff>
else:
    <do the other stuff>
```

File: robottelo/api/utils.py
Comment (mshriver) [inline robottelo/api/utils.py:973]: Consider changing how this function works, to make it more usable with tests and parametrization.  If the function took a dictionary of settings, then tests could actually use this function to change settings relevant to test data or cases - instead of just statically configuring whatever is in the properties file.

You can update the passed dictionary with values from the properties files for anything that wasn't passed in, retaining your default behavior here.

File: robottelo/api/utils.py
Comment (mshriver) [inline robottelo/api/utils.py:988]: Can use an f-string here too :)

File: robottelo/rhsso_utils.py
Comment (mshriver) [reply robottelo/rhsso_utils.py:167]: @JacobCallahan Thoughts here, especially with future use of Broker's Host class and SSH interface?

File: robottelo/api/utils.py
Comment (omkarkhatavkar) [reply robottelo/api/utils.py:973]: If you think strictly this needs to be done in this PR  I will do it. As this is old code, I will have te refactor everything can take more time. If this PR has already over more than 30 days. Based on what you have suggested can be done as part of another issue, unless you do not agree.

File: robottelo/rhsso_utils.py
Comment (mshriver) [inline robottelo/rhsso_utils.py:168]: Its overkill to use f-strings here, with a single value. You can just reference the name directly, its already a string. If you're concerned about it being a different type, you can just use `str(settings.server.ssh_key)`

File: robottelo/rhsso_utils.py
Comment (mshriver) [inline robottelo/rhsso_utils.py:135]: These functions have limited use when they're statically returning a string formatted with data from settings. Consider taking hostname and realm as kwargs (for this function, other kwargs for the other similar functions), and constructing the strings from either the provided kwargs or values from settings.

File: robottelo/api/utils.py
Comment (mshriver) [reply robottelo/api/utils.py:973]: Ah, I see now this was just moved out of a test module. Agreed this could be enhanced later, since you didn't write it specifically for this PR.

Comment (omkarkhatavkar) [reply robottelo/api/utils.py:973]: create issue here https://github.com/SatelliteQE/robottelo/issues/7990 will work on future

File: robottelo/rhsso_utils.py
Comment (omkarkhatavkar) [reply robottelo/rhsso_utils.py:135]: Hence this is dependant of https://github.com/SatelliteQE/robottelo/issues/7990 issue will work on later sprint

File: robottelo/rhsso_utils.py
Comment (omkarkhatavkar) [reply robottelo/rhsso_utils.py:168]: I think casting to str is also not needed as there is nothing an integer

**Review by mshriver — APPROVED** [review]

File: robottelo/rhsso_utils.py
Comment (omkarkhatavkar) [reply robottelo/rhsso_utils.py:167]: @JacobCallahan Do you have anything to add here?

Comment (JacobCallahan) [reply robottelo/rhsso_utils.py:167]: @omkarkhatavkar once the broker integration is merged, you can use its ssh class, which uses ssh2-python and is significantly faster than paramiko.

Comment (omkarkhatavkar) [reply robottelo/rhsso_utils.py:167]: I added https://github.com/SatelliteQE/robottelo/pull/7925#discussion_r474391407 my comment here that there was no performance issue here, Functionally paramiko was not working for me hence used expect. Holding this PR for that change does not make any sense to me. We are already near to 6.8 GA and I'm still waiting for my PR's to get merged.

Comment (mshriver) [reply robottelo/rhsso_utils.py:167]: @omkarkhatavkar I don't think @JacobCallahan is suggesting this PR be held for this. Simply using this thread to discuss what changes could come in the future.

Comment (mshriver) [reply robottelo/rhsso_utils.py:167]: I've approved, just waiting on re-review/merge from @mirekdlugosz or review/merge from @SatelliteQE/robottelo-tier-2-reviewers

Comment (omkarkhatavkar) [reply robottelo/rhsso_utils.py:167]: Okay, Thanks  @mshriver and @JacobCallahan for clarification. I will see if  I can use ssh2-python to solve this issue later for faster performance.

**Review by mirekdlugosz — APPROVED** [review]
ACK
I see all comments are resolved, merging.

**Pull request merged**
